Adding new analysis TRestTrackAlphaAnalysisProcess

juanangp opened the pull request:
![juanangp](https://badgen.net/badge/PR%20submitted%20by%3A/juanangp/blue) ![Large: 1974](https://badgen.net/badge/PR%20Size/Large%3A%201974/red) [![](https://gitlab.cern.ch/rest-for-physics/tracklib/badges/alphaAna/pipeline.svg)](https://gitlab.cern.ch/rest-for-physics/tracklib/-/commits/alphaAna) [![](https://gitlab.cern.ch/rest-for-physics/framework/badges/alphaAna/pipeline.svg)](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/alphaAna)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

New analysis have been added to tracklib to analyze alpha tracks, which was developed for AlphaCAMM.
A macro has been added to display the analysis results.

Summary of latest changes as requested in this PR:

- New `TRestTrackViewerProcess` for track visualization.
- New drawing methods have been added to `TRestTrackEvent`
- Renaming `TRestTrackAlphaAnalysisProcess` to `TRestTrackLineAnalysisProcess`
- New generic class `TRestTrackLinearizationProcess`, which is meant for track linearization.
- Adding new methods to `TRestTrackReductionProcess` and `TRestTrackPathMinimizationProcess`
- Now the macro lies in a dedicated repository for AlphaCAMM.

jgalan commented on macros/REST_Track_PlotAlphaAna.C:
```diff
@@ -0,0 +1,457 @@
+
+}
+
+TPad *DrawHits ( TRestDetectorHitsEvent *hitsEvent, const TVector3 &origin, const TVector3 &end, TVector3 &max, TVector3 &min, TVector3& nBins) {
```
It would be good that Drawing methods would be common inside `TRestTrackEvent`

jgalan commented on macros/REST_Track_PlotAlphaAna.C:
```diff
@@ -0,0 +1,457 @@
+const int offset=10;//bins
+
+
+TPad *DrawTrack ( TRestTrackEvent *trackEvent, const TVector3 &origin, const TVector3 &end, const TVector3 &max, const TVector3 &min, const TVector3& nBins) {
```
Perhaps better to implement at `TRestTrackEvent`

jgalan replied:
I understand, but anyway it could be implemented at TRestTrackEvent and the macro would just call to this method?

See for example `SetBoundaries` method at `TRestDetectorHitsEvent`.

Of course, those methods could be even further promoted to a higher class, it would be good to have dedicated drawing classes for hits and for time signals that centralize that kind of staff.

jgalan commented on src/TRestTrackAlphaAnalysisProcess.cxx:
```diff
@@ -0,0 +1,281 @@
+///
+///             TRestTrackAlphaAnalysisProcess.cxx
+///
+///_______________________________________________________________________________
```
Please, add the official headers. See for example `TRestGeant4Metadata` style. Add also some documentation, include figures if possible.

jgalan commented on src/TRestTrackAlphaAnalysisProcess.cxx:
```diff
@@ -0,0 +1,281 @@
+
+    debug<<"Track length "<<length<<" angle: "<<angle<<endl;
+
+    // A new value for each observable is added
```
Document observables in documentation

jgalan commented on src/TRestTrackAlphaAnalysisProcess.cxx:
```diff
@@ -0,0 +1,281 @@
+
+//______________________________________________________________________________
+void TRestTrackAlphaAnalysisProcess::InitFromConfigFile() {
+  fTrackBalance = StringToDouble(GetParameter("trackBalance", "0.65"));
```
There is no need to implement this method. Simply remove completely `InitFromConfigFile` from `TRestTrackAlphaAnalysisProcess`.

Then, `TRestEventProcess::InitFromConfigFile` will be responsible to identify a `fDummyVariable` with the parameter given in the config file `dummyVariable`.

In the case of `trackBalance` you just need that a data member named `fTrackBalance` exists.

The declaratio at the header would look as:

```
/// A parameter to define the track balance ... bla bla bla
Double_t fTrackBalance = 0.65;
```

nkx111 replied:
Actually the methods `LoadDefaultConfig()`, `LoadConfig()`, `TRestTrackAlphaAnalysisProcess(char* cfgfile)` are not needed either.

jgalan commented on inc/TRestTrackAlphaAnalysisProcess.h:
```diff
@@ -0,0 +1,82 @@
+
+   protected:
+    // add here the members of your event process
+
```
Methods to be documented at CXX.

jgalan commented on inc/TRestTrackAlphaAnalysisProcess.h:
```diff
@@ -0,0 +1,82 @@
+///
+///             TRestTrackAlphaAnalysisProcess.h
+///
+///_______________________________________________________________________________
```
Use official headers

jgalan commented:
> This PR is somehow related with [rest-for-physics/framework#31](https://github.com/rest-for-physics/framework/pull/31)

Not sure how it is related to #31, perhaps you refer to the wrong PR?

juanangp commented on macros/REST_Track_PlotAlphaAna.C:
```diff
@@ -0,0 +1,457 @@
+
+}
+
+TPad *DrawPulses ( TRestRawSignalEvent *rawSignalEvent, TRestDetectorReadout *fReadout) {
```
The problem is that I need the readout info to distinguish between X and Y tracks and it is not straithforward to get it.

jgalan replied:
The hits at `TRestDetectorSignalEvent` now about the readout, and also are capable to identify if they are X, Y type.

juanangp replied:
These are TRestRawSignalEvents, I have already implemented a method to plot a vector of sIDs.

jgalan replied:
I thought some pulse drawing methods at the detector level were alreaqdy implemented by you at `TRestDetectorReadoutEventViewer`

jgalan commented on macros/REST_Track_PlotAlphaAna.C:
```diff
@@ -0,0 +1,341 @@
+#include <TRestRawSignalEvent.h>
+
+#include "dirent.h"
+
```
This method should not be implemented at the level of `tracklib`. If REST is not compiled with `detectorlib` or `rawlib` there will be problems when this macro is being loaded in `restRoot`.

juanangp replied:
Indeed, you are rigth.

Do you prefer to move to general macros directory? I don't know what is the best strategy here.

jgalan replied:
The libraries design is to focus on independent development of libraries. So that a particular library keeps a minimum of dependencies with other libraries. I am sure it is possible to think a way to keep that strategy alive.

In the case this is not possible, and using `rawlib` routines inside the `tracklib` is a must (although I would prefer to find a way to avoid) then the code instantiating `rawlib` must be encapsulated using a compilation flag, that could be created for example at `cmake/MacroRootDict.cmake` at `COMPILEDIR`, and define a `#ifdef RESTLIB_RAW`, not sure if this already exist. @nkx?

However, I would prefer that such a possibility does not exist, the present structure is the result of an effort to organise the code project, and I believe keeping isolated developments at the libraries as they are is the best to avoid future chaos. I understand this requires an effort, but I am open to discuss about possible solutions for the present implementation.

Keep present that if you add your macro to the `tracklib` pipeline, where only track routines are tested, the execution will fail.

https://gitlab.cern.ch/rest-for-physics/tracklib/-/pipelines/3637513

At the rawlib there are already a few tests running that do not need other libraries, we should push for something similar at tracklib

https://gitlab.cern.ch/rest-for-physics/tracklib/-/pipelines/3637513

jgalan replied:
Adding it to the framework domain is also not desirable. The framework should know nothing about signal processing, physical track event reconstruction, detectors, etc. It should only implement basic analysis routines not connected to those domains. Such as fitting analysis tree observables, etc. This is discussed in the paper.

nkx111 replied:
To add dependency we can just call:
```
set(deps detector geant4 track raw)
COMPILELIB(deps)
```
Currently all libraries except connectorslib have no dependency on others. So it would be strange to make tracklib dependent on rawlib.

jgalan commented on macros/REST_Track_PlotAlphaAna.C:
```diff
@@ -0,0 +1,341 @@
+ return fPad;
+
+}
+
```
Should probably need to be implemented at `TRestTrackEvent`.

jgalan commented on macros/REST_Track_PlotAlphaAna.C:
```diff
@@ -0,0 +1,341 @@
+  endYZGr->Draw("LP");
+
+}
+
```
Should be implemented at `TRestTools`? There is no documentation, what it is doing this method? Perhaps there is already a similar method at `TRestTools` doing what is doing this one?

juanangp commented:
I have implemented all the drawing classes in `TRestTrackEvent` and I have removed the macro, summary of the changes:
- New `TRestTrackViewerProcess` for track visualization.
- New drawing methods have been added to `TRestTrackEvent`
- Renaming `TRestTrackAlphaAnalysisProcess` to `TRestTrackLineAnalysisProcess`
- New class `TRestTrackLinearizationProcess`
- Adding new methods to `TRestTrackReductionProcess` and `TRestTrackPathMinimizationProcess`
- Now the macro lies in a dedicated repository for AlphaCAMM.

jgalan commented:
> I have implemented all the drawing classes in `TRestTrackEvent` and I have removed the macro, summary of the changes:
> 
> * New `TRestTrackViewerProcess` for track visualization.
> * New drawing methods have been added to `TRestTrackEvent`
> * Renaming `TRestTrackAlphaAnalysisProcess` to `TRestTrackLineAnalysisProcess`
> * New class `TRestTrackLinearizationProcess`
> * Adding new methods to `TRestTrackReductionProcess` and `TRestTrackPathMinimizationProcess`
> * Now the macro lies in a dedicated repository for AlphaCAMM.

This is probably to be added at the description of the PR. 

I will have a look to the PR, but later.

jgalan commented on inc/TRestTrackLineAnalysisProcess.h:
```diff
@@ -0,0 +1,71 @@
+
+#include "TRestEventProcess.h"
+
+//! A process to analyze alpha tracks
```
//! A process to identify and extract the track ends in linear approximation?

jgalan commented on inc/TRestTrackLineAnalysisProcess.h:
```diff
@@ -0,0 +1,71 @@
+    void Initialize();
+
+   protected:
+    /// A parameter to define the track balance, otherwise the event is rejected
```
not clear, when the event is rejected?

jgalan commented on inc/TRestTrackLinearizationProcess.h:
```diff
@@ -0,0 +1,84 @@
+        EndPrintProcess();
+    }
+
+    // This function performs the track linearization towards a given number of nodes
```
This description should be written in the implementation, at the cxx file.

jgalan commented on src/TRestTrackLineAnalysisProcess.cxx:
```diff
@@ -0,0 +1,252 @@
+
+ClassImp(TRestTrackLineAnalysisProcess);
+
+//______________________________________________________________________________
```
Not the good formatting. Using `REST_MakeProcess` will produce the standard formatting.

jgalan commented on src/TRestTrackLineAnalysisProcess.cxx:
```diff
@@ -0,0 +1,252 @@
+//////////////////////////////////////////////////////////////////////////
+/// The TRestTrackLineAnalysis process is meant to analyze lienar tracks,
+/// such as alpha tracks. It should retreive the information after performing
+/// TRestTrackLinearization process, for the time being only works when
```
What it happens if `TRestTrackLinearizationProcess` was not used earlier?

juanangp replied:
In fact, is not required to use `TRestTrackLinearizationProcess` before. It should work as soon as the reduced hits are ordered. I am going to change the description.

jgalan replied:
Ok. It is this process doing a linear fit to the hits? In that case the hit order would also be not relevant.

juanangp replied:
The hit order is required in the sense that it should follow the minimum path, this is guaranteed after performing `TRestTrackLinearizationProcess` . It can be also reached using `TRestTrackPathMinimizationProcess` after a `TRestTrackReductionProcess`.

However, at this stage the hits ordering doesn't require to have any preferred direction e.g. a hit ordering of `1->2->3->4->5` should give same solution as an inverted ordering `5->4->3->2->1`

jgalan replied:
I see, we should probably introduce a mechanism that allows us to do something like `ProcessRequired("TRestTrackPathMinimizationProcess")` so that if the process does not exist it will be returning an error.

Another potential issue, perhaps @nkx111 sees an easy way

jgalan replied:
I created an issue about it rest-for-physics/framework#182

jgalan commented on src/TRestTrackLineAnalysisProcess.cxx:
```diff
@@ -0,0 +1,252 @@
+/// such as alpha tracks. It should retreive the information after performing
+/// TRestTrackLinearization process, for the time being only works when
+/// XZ and YZ are present in the readout. Moreover, the hits have to be stored
+/// following the minimum path. It assumes that only one alpha track
```
Probably it could be generalised to a multi track event by picking up the track with the maximum energy reusing one of these methods:

```
    TRestTrack* GetMaxEnergyTrackInX();
    TRestTrack* GetMaxEnergyTrackInY();

    TRestTrack* GetMaxEnergyTrack(TString option = "");
```

juanangp replied:
Done

jgalan commented on src/TRestTrackLineAnalysisProcess.cxx:
```diff
@@ -0,0 +1,252 @@
+/// following the minimum path. It assumes that only one alpha track
+/// is present, using the parameter fTrackBalance to define the minimun energy
+/// fTrackBalance*totalEnergy of the total XZ or YZ energy to assume a single
+/// track event, if the condition is not fulfilled the event is rejected.
```
I think we should try to do not reject events. If it is not a good event, then we will be able to discriminate it later using filters at the `analysisTree`.

juanangp replied:
Ok, the parameted `fTrackBalance` has been removed, now we have 2 observables instead `trackBalanceX` and `trackBalanceY`

jgalan commented on src/TRestTrackLinearizationProcess.cxx:
```diff
@@ -0,0 +1,230 @@
+//////////////////////////////////////////////////////////////////////////
+/// The TRestTrackLinearization process is meant to analyze linear tracks,
+/// such as alpha tracks. It should retreive the information after performing
+/// TRestTrackLinearization process, not implemented for XYZ tracks.
```
`GetHitsProjection` was implemented time ago to reconstruct the profile a physical track event along the track. The track could be a complex electron track event (no need to be linear). I understand this process is doing now something else, I could enter in the details of how this is being used here, but I dont see how this is required to perform the fitting to the raw hits.

The process producing a track profile could be probably named `TRestTrackProfileProcess`, generating a std::vector or a TGraph with N points where the hits energy have been projected.

juanangp replied:
Sorry, I don't understand this comment. Do you want to rename `TRestTrackLinearizationProcess` to `TRestTrackProfileProcess`?

Perhaps the naming of `GetHitsProjection` function is not adequate? I can rename the function. This function is actually performing the track linearization.

Or you want to split the process in two different processes one doing the linear fit or profile and another process doing the average of the closest hits? Also, I don't understand why do you want to store the results as an `std::vector` or `TGraph` since I believe is better to store as `TRestVolumeHits`

jgalan replied:
We should discuss this online

jgalan commented on src/TRestTrackPathMinimizationProcess.cxx:
```diff
@@ -79,226 +56,313 @@ TRestEvent* TRestTrackPathMinimizationProcess::ProcessEvent(TRestEvent* evInput)
+
+// Return the index with the shortest path solving Travelling Salesman Problem (TSP) using
+// brute force, algorithm complexity is n!, the minimum path is guarantee
+//______________________________________________________________________________
```
This is not the standard format for method documentation

jgalan commented on src/TRestTrackViewerProcess.cxx:
```diff
@@ -0,0 +1,189 @@
+///
+void TRestTrackViewerProcess::DrawOriginEnd(TPad* pad, std::vector<TGraph*>& origin,
+                                            std::vector<TGraph*>& end, std::vector<TLegend*>& leg) {
+    for (int i = 0; i < 2; i++) {
```
It should check if `origin` and `end` got exactly 2-elements. I would allow to do not check this if the method were `private` and it is clear that the only case is that it will contain 2-elements. But the method was declared `public`.

juanangp replied:
Done

jgalan commented on src/TRestTrackViewerProcess.cxx:
```diff
@@ -0,0 +1,189 @@
+/// \brief Get origin and end of the track if trackLineAnalysis have
+/// been performed
+///
+void TRestTrackViewerProcess::GetOriginEnd(TRestAnalysisTree* anaTree, std::vector<TGraph*>& origin,
```
Why it receives a pointer to `TRestAnalysisTree`? The process will always have access to the analysis tree, i.e. using `this->GetFullAnalysisTree()` or `this->GetAnalysisTree()`.

jgalan replied:
Perhaps this is a method to be implemented at `TRestTrackEvent`, or event at a new class `TRestTrackPainter` that implements statics methods, then `TRestTrackEvent` could inherit from `TRestTrackPainter`.

juanangp replied:
Note that the function is declared `static`, so it need a `TRestAnalysisTree` passed to the function. I prefer to use `static` functions in this case because it allows me to use elsewhere.

jgalan replied:
I understand the benefit of using static functions, just that perhaps should be placed at a common class contextualizing it, such as `TRestTrackPainter`. For me, it makes sense to associate a `GetOriginEnd` to a `TrackEvent`, but not to the process.

jgalan replied:
Or perhaps a `TRestTrackAnalysis` that contains a `GetOriginEnd` implementing an algorithm that extracts origin/end from the track, but not from the analysis tree. I would do the other way around. I would use the method to fill the analysis tree.

juanangp replied:
I agree with you that most of the static methods should be implemented in a dedicated `class`, perhaps in this case would be better to use a `namespace`.

I like the idea to have different `namespace` such as `TRestTrackPainter` and `TRestTrackAnalysis`. However, I think that this approach should be more generic:
 - `TRestTrackPainter` should implement all the track drawing methods.
 - `TRestTrackAnalysis` should implement all the track analysis methods.

We should probably use a similar schema for all the libraries by defining dedicated `TRestXXXPainter` and `TRestXXXAnalysis` namespaces in every single library.

However, this approach would require several changes in the code that cannot be addressed in this PR. I can give a hand with the migration, but it would require a lot of work. Let me know what do you think.

jgalan replied:
Yes, that sounds good. If someone is leading that, I will also give a hand with the migration. 

For this PR I think we need to chat online about the real need to request analysisTree variables in that method.

juanangp replied:
I am working on that, `TRestTrackViewerProcess` can call a static function on `TRestTrackLineAnalysisProcess` to calculate the origin and the end of the track.

Later on we can migrate all the static functions to `TRestTrackPainter` and `TRestTrackAnalysis`, or do yo prefer to do in this PR just with few static functions mentioned before?

jgalan replied:
Yes, that could be done later.

juanangp replied:
Ok, I have created an issue about this topic

https://github.com/rest-for-physics/framework/issues/183

jgalan commented on src/TRestTrackViewerProcess.cxx:
```diff
@@ -0,0 +1,189 @@
+    std::vector<std::string> eName = {"endX", "endY", "endZ"};
+
+    for (int i = 0; i < 3; i++) {
+        std::string fName = "alphaTrackAna_" + oName[i];
```
`alphaTrackAna_` is defined by the user at the RML, then the name could be anything else, isnt?

jgalan replied:
If this observable was extracted from `TRestTrackAnalysisProcess` one could perhaps retrieve the process name as follows:

```
fRunInfo->GetMetadataClass("TRestTrackAnalysisProcess")->GetName();
```

However, it might happen that there are two classes `TRestTrackAnalysisProcess` inside the data processing chain.

Not sure if there is a better way @nkx?

juanangp replied:
I can see that `GetObservableValue` is used somewhere else in a similar way.

![image](https://user-images.githubusercontent.com/80903717/163877981-68411cfb-1f19-4828-9433-b9937e95e183.png)

I can think in some analysis processes that may use some variables from the analysis tree as input. So, I don't know what is the best solution here. As far as I understand the observables are always retreived from the `rml`. 

Perhaps we should create an issue about this?

jgalan replied:
Yes, I dont think it is good to hardcode the process name,  since the user might change that value. On top of that, as I said earlier, it might happen that exist two `TRestRawSignalAnalysisProcess` inside the data processing chain. Which of them `TRestDetectorSingleChannelAnalysisProcess` would use?

On the other hand, the process `TRestDetectorSingleChannelAnalysisProcess` uses a `TRestDetectorSignalEvent`, so I believe the `max_amplitude`, `thr_integral` and `PeakAmplitudeIntegral` maps could be calculated directly inside the process without the need to retrieve them from a previous analysis. Isn't?

jgalan replied:
Just created issue rest-for-physics/detectorlib#47

juanangp commented on src/TRestTrackLinearizationProcess.cxx:
```diff
@@ -0,0 +1,230 @@
+//______________________________________________________________________________
+TRestEvent* TRestTrackLinearizationProcess::ProcessEvent(TRestEvent* evInput) {
+    fTrackEvent = (TRestTrackEvent*)evInput;
+    fOutTrackEvent->SetEventInfo(fTrackEvent);
```
`SetEventInfo` has been removed

jgalan commented on src/TRestTrackLineAnalysisProcess.cxx:
```diff
@@ -132,48 +132,50 @@ TRestEvent* TRestTrackLineAnalysisProcess::ProcessEvent(TRestEvent* evInput) {
+        SetObservableValue("downwards", downwards);
+        SetObservableValue("totalEnergy", trackEnergy);
+    } else {
+        return nullptr;
```
So, I would define the values of the variables associated with `SetObservableValue` to dummy values, and update them in case `tckX && tckY == true`.

juanangp replied:
Indeed, I am going to change it.

jgalan commented on src/TRestTrackLinearizationProcess.cxx:
```diff
@@ -103,7 +103,7 @@ TRestEvent* TRestTrackLinearizationProcess::ProcessEvent(TRestEvent* evInput) {
 
     for (int t = 0; t < fTrackEvent->GetNumberOfTracks(); t++) {
-        if (!fTrackEvent->isTopLevel(t)) continue;
+        if (fTrackEvent->GetLevel(t) != 1) continue;
```
Why `!= 1`? Should not be` > 0`? If a track is level 2 it will not continue.

juanangp replied:
This is the way that I found to make it work, it seems that the lowest track level is 1 instead of 0. If you want I can change the statement to `if (fTrackEvent->GetLevel(t) > 1) continue;`

jgalan replied:
Yes, I think it should be `>` in case in the future another process plugged just before increases the number of levels. This way you are sure the processing only affects to the lower track level.

jgalan replied:
Ok, I see `!=` is also valid, but for some reason for me it is easier to understand because I have to check mentally the behaviour only at the right of 1.

jgalan commented:
Please, before merging increase the library version, tag it and add the release notes.
